pkg/idtools/idtools_unix.go

```diff
@@ -60,9 +71,17 @@ func mkdirAs(path string, mode os.FileMode, ownerUID, ownerGID int, mkAll, chown
 	// even if it existed, we will chown the requested path + any subpaths that
 	// didn't exist when we called MkdirAll
 	for _, pathComponent := range paths {
-		if err := os.Chown(pathComponent, ownerUID, ownerGID); err != nil {
+		// Normally a Chown is a no-op if uid/gid match, but in some cases this can still cause an error, e.g. if the
+		// dir is on an NFS share, so don't call chown unless we absolutely must.
+		stat, err := system.Stat(pathComponent)
+		if err != nil {
 			return err
 		}
+		if stat.UID() != uint32(ownerUID) || stat.GID() != uint32(ownerGID) {
```

Comment:
Another option could be to implement a own `Chown` (or with a similar name) under `system` so that other places could benefit using a more "safe" version of chown.
Also there are cases where the chown is even repeated (for other uid, gid), like

https://github.com/cpuguy83/docker/blob/32a6db4c971910aefca04b1c9609a19549ca0342/daemon/container_operations_unix.go#L307

You might say, that kind of files is not usually on something like nfs so no problem there but to get a more consistent behavior in my opinion a better solution is to have the check at `chown` level directly

Reply:
I was a little averse to implementing our own chown, but since we're already importing system here and also have this pattern elsewhere, it's likely an good thing to do.

Reply:
Even a `lazyChown()` that isn't exported could be used here and above instead of checking uid/gid twice. It would also remove the duplicate comment.